crates/subgraph/src/apy.rs

```diff
@@ -0,0 +1,777 @@
+#[serde(rename_all = "camelCase")]
+#[typeshare]
+pub struct OrderAPY {
+    pub order_id: String,
+    pub order_hash: String,
+    pub apy: Option<DenominatedAPY>,
+    #[typeshare(typescript(type = "number"))]
+    pub start_time: u64,
+    #[typeshare(typescript(type = "number"))]
+    pub end_time: u64,
+    pub inputs_token_vault_apy: Vec<TokenVaultAPY>,
+    pub outputs_token_vault_apy: Vec<TokenVaultAPY>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct TokenPair {
+    input: Erc20,
+    output: Erc20,
+}
+
+/// Given an order and its trades and optionally a timeframe, will calculates
+/// the APY for each of the entire order and for each of its vaults
+pub fn get_order_apy(
+    order: Order,
+    trades: &[Trade],
+    start_timestamp: Option<u64>,
+    end_timestamp: Option<u64>,
+) -> Result<OrderAPY, OrderbookSubgraphClientError> {
+    let one = I256::from_str(ONE).unwrap();
+    if trades.is_empty() {
+        return Ok(OrderAPY {
+            order_id: order.id.0.clone(),
+            order_hash: order.order_hash.0.clone(),
+            start_time: start_timestamp.unwrap_or(0),
+            end_time: end_timestamp.unwrap_or(chrono::Utc::now().timestamp() as u64),
+            inputs_token_vault_apy: vec![],
+            outputs_token_vault_apy: vec![],
+            apy: None,
+        });
+    }
+    let vols = get_vaults_vol(trades)?;
+    let token_vaults_apy = get_token_vaults_apy(trades, &vols, start_timestamp, end_timestamp)?;
+
+    // build an OrderApy struct
+    let mut start_time = u64::MAX;
+    let mut end_time = 0_u64;
+    let mut inputs: Vec<TokenVaultAPY> = vec![];
+    let mut outputs: Vec<TokenVaultAPY> = vec![];
+    for item in &token_vaults_apy {
+        if item.start_time < start_time {
+            start_time = item.start_time;
+        }
+        if item.end_time > end_time {
+            end_time = item.end_time;
+        }
+        if order
+            .inputs
+            .iter()
+            .any(|v| v.vault_id.0 == item.id && v.token == item.token)
+        {
+            inputs.push(item.clone());
+        }
+        if order
+            .outputs
+            .iter()
+            .any(|v| v.vault_id.0 == item.id && v.token == item.token)
+        {
+            outputs.push(item.clone());
+        }
+    }
+    let mut order_apy = OrderAPY {
+        order_id: order.id.0.clone(),
+        order_hash: order.order_hash.0.clone(),
+        start_time,
+        end_time,
+        inputs_token_vault_apy: inputs,
+        outputs_token_vault_apy: outputs,
+        apy: None,
+    };
+
+    // get pairs ratios
+    let pair_ratio_map = get_pairs_ratio(&order_apy, trades);
+
+    // try to calculate all vaults capital and volume denominated into any of
+    // the order's tokens by checking if there is direct ratio between the tokens,
+    // multi path ratios are ignored currently and results in None for the APY.
+    // if there is a success for any of the denomination tokens, checks if it is
```

Comment:
most positive compared to what? asking because, fater all conversions, they all represent the same value?
lets say the total vol converted to usdt is 100, and total vol converted to wbnb is 0.2, and the ratio of wbnb/usdt is 500, so these 2 vols are the same thing, same value, same for total starting captial converted to either of usdt or wbnb as well, so the final apy will be the same if you mul it by ratio to convert to one another.
ie the APY being + or - wont be affected by the denominating token at all, if the apy is negative, it is negative no matter what token denomination is chosen, same for positive apy, only the apy variable will change, but the underlying value will be the same for all denominations when the ratio of tokens si factored in to convert from one denomination to other.

Reply:
most postive in the unconverted form

Reply:
<img width="1266" alt="Screenshot 2024-10-20 at 14 52 29" src="https://github.com/user-attachments/assets/92911e44-ac6f-448c-a6f1-45878c53aace">

for example, here we are positive in QUICK (old) so that's the denomination for our revenue, and so our positive APY

we're negative in new QUICK currently, so it's not the denomination of our profit